City-of-Helsinki/yjdh
# Hl 655 add s3

## Description :sparkles:
This PR add S3 file system support in anticipation of the upcoming staging / production environments.
As the current dev/test environments use Azure Blob storage, the S3 support is optionally enable by setting the 
`USE_S3` env-variable  to True and then adding  the  S3-specific connection variables:

- `S3_ENDPOINT_URL`
- `S3_ACCESS_KEY_ID`
- `S3_SECRET_ACCESS_KEY`
- `S3_STORAGE_BUCKET_NAME`

The PR also adds a [MinIO](https://min.io/) -container that can be used to test the S3-storage in the local development environment.

## Issues :bug:
MinIO cannot keep up with the amount of mock files generated by the test factories so running the `python manage.py seed `and `pytest` commands takes a looong time or fails.  So before running backend tests or the seed command, switch back to the local disk as a default storage by setting `USE_S3=0`. This is clumsy, and a possible workaround could be to set the `USE_S3 `to False when running tests like this ` USE_S3=False pytest . `.

## Testing :alembic:

0. Enable the S3 storage by settting the necessary variables from `.env.benefit-backend.example`.
1. Build the containers
2. Visit the minIO console at the `http://localhost:9090`
3. Login with username `minio-root` and password `minio-pass`
4. Navigate to Object Browser and select `local-s3-bucket`
5. From the applicant UI, upload some attachments
6. From the minIO console, click refresh and verify that the files uploaded from the UI are present

To confirm that a file is stored from the backend:
1.  Run` python manage.py shell`
2. `from django.core.files.storage import default_storage`
3. `file = default_storage.open('testfile', 'w')`
4. `file.write('test text')`
5. `file.close()`
6.  Refresh the bucket in the minIO console and verify that testfile was uploaded.

## Screenshots :camera_flash:
<img width="1437" alt="Screenshot 2023-02-20 at 9 46 32" src="https://user-images.githubusercontent.com/33894149/220060801-cb1089bc-2876-49a6-86b5-18b756fc7f99.png">
<img width="1256" alt="Screenshot 2023-02-20 at 10 21 23" src="https://user-images.githubusercontent.com/33894149/220060829-c0235417-cf2f-4573-b77a-46ac008f1cb1.png">

<img width="1395" alt="Screenshot 2023-02-20 at 10 21 35" src="https://user-images.githubusercontent.com/33894149/220060870-3e17c0c7-af5b-40d8-ac53-4f7aefe82d8a.png">

## Additional notes :spiral_notepad:
